lib/webhdfs/client_v1.rb

```diff
@@ -60,6 +61,104 @@ def initialize(host='localhost', port=50070, username=nil, doas=nil, proxy_addre
       @kerberos = false
       @kerberos_keytab = nil
       @http_headers = http_headers
+
+      @jmx = nil
+    end
+
+    def self.simple(opts = {})
+      opts = {
+        :jmx => ENV['JMX'],
+        :host => ENV['DEFAULT_NAMENODE'] || 'localhost',
+        :port => 50070,
+        :kerberos => ENV['KERBEROS'] && true,
+        :kerberos_keytab => ENV['KERBEROS_KEYTAB'],
+
+        :retry_known_errors => false,
+        :retry_times => 1,
+        :retry_interval => 1,
+
+        :httpfs_mode => false,
+
+        :ssl => false,
+      }.merge(opts)
+
+      client = new
+
+      client.host = opts[:host]
+      client.port = opts[:port]
+      client.username = opts[:username]
+
+      client.doas = opts[:doas]
+
+      client.proxy_address = opts[:proxy_address]
+      client.proxy_port = opts[:proxy_port]
+
+      client.retry_known_errors = opts[:retry_known_errors]
+      client.retry_times = opts[:retry_times]
+      client.retry_interval = opts[:retry_interval]
+
+      client.httpfs_mode = opts[:httpfs_mode]
+
+      client.ssl = opts[:ssl]
+      client.ssl_ca_file = opts[:ssl_ca_file]
+      client.ssl_cert = opts[:ssl_cert]
+      client.ssl_key = opts[:ssl_key]
+      client.ssl_version = opts[:ssl_version]
+
+      client.http_headers = opts[:http_headers]
+
+      client.kerberos = opts[:kerberos]
+      client.kerberos_keytab = opts[:kerberos_keytab]
+
+      client.jmx = opts[:jmx]
+
+      client
+    end
+
+    def self.keytab_path_set?
+      if ENV['KEYTAB_PATH'].nil? || ENV['KEYTAB_PATH'].empty?
+        WebHDFS.logger.fatal("The kerberos keytab is not set")
+        false
+      else
+        true
+      end
+    end
+
+    def ensure_keytab_path_set
+      raise WebHDFS::KerberosError, "The kerberos keytab must be set" unless keytab_path_set?
+    end
+
+    def operational?
+      !!stat('/')
+    rescue StandardError => e
+      WebHDFS.logger.error("Failed to access HDFS with error #{$!}")
+      if e.is_a?(NameError) && e.message =~ /undefined local variable or method `min_stat'/
```

Comment:
There are a lot of pretty specific error handlers like this around -- just curious how you came up with all these? was this just trial and error from running into them? Or were they all documented somewhere?

Reply:
This was part of the code that was ported over from what was duplicated across various other Factual projects. This one is from https://github.com/Factual/nim/blob/a0470e2bf6a5cd15a37a269be5e2c03142cc018c/lib/hdfs_client.rb#L40-L49